File: app/src/main/java/com/krsk/qiitareader/presentation/presenter/ItemListPresenter.java
package com.krsk.qiitareader.presentation.presenter;

import android.app.Fragment;
import android.net.Uri;
import android.support.customtabs.CustomTabsIntent;

import com.krsk.qiitareader.domain.interactor.UseCase;
import com.krsk.qiitareader.domain.model.entity.Item;
import com.krsk.qiitareader.presentation.fragment.ItemListFragment;
import com.krsk.qiitareader.presentation.internal.di.PerFragment;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Named;

/**
 * Created by bl-lia on 2/2/16.
 */
@PerFragment
public class ItemListPresenter implements Presenter {

    private final ItemListFragment fragment;
    private final UseCase<List<Item>> getItemsUseCase;

    @Inject
    public ItemListPresenter(Fragment fragment,
                             @Named("getItems") UseCase<List<Item>> getItemsUseCase) {
        this.fragment = (ItemListFragment) fragment;
        this.getItemsUseCase = getItemsUseCase;
    }

    @Override
    public void resume() {

    }

    @Override
    public void pause() {

    }

    @Override
    public void destroy() {

    }

    public void listItems() {
        this.getItemsUseCase.execute(new UseCase.UseCaseSubscriber<List<Item>>() {

            private List<Item> items = new ArrayList<>();

            @Override
            public void onCompleted() {
                fragment.resetList(this.items);
            }

            @Override
            public void onNext(List<Item> items) {
                this.items.addAll(items);
            }
        });
    }

    public void navigateItemDetail(Item item) {
        final CustomTabsIntent intent = new CustomTabsIntent.Builder()
                                                            .setShowTitle(true)
                                                            .build();
        intent.launchUrl(fragment.getActivity(), Uri.parse(item.url));
    }

}

File: app/src/main/kotlin/com/krsk/qiitareader/presentation/presenter/ItemListPresenter.kt
package com.krsk.qiitareader.presentation.presenter

import android.app.Fragment
import android.net.Uri
import android.support.customtabs.CustomTabsIntent

import com.krsk.qiitareader.domain.interactor.UseCase
import com.krsk.qiitareader.domain.model.entity.Item
import com.krsk.qiitareader.presentation.fragment.ItemListFragment
import com.krsk.qiitareader.presentation.internal.di.PerFragment

import java.util.ArrayList

import javax.inject.Inject
import javax.inject.Named

/**
 * Created by bl-lia on 2/2/16.
 */
@PerFragment
class ItemListPresenter
@Inject
constructor(fragment: Fragment,
            @Named("getItems") private val getItemsUseCase: UseCase<List<Item>>) : Presenter {

    private val fragment: ItemListFragment

    init {
        this.fragment = fragment as ItemListFragment
    }

    override fun resume() {

    }

    override fun pause() {

    }

    override fun destroy() {

    }

    fun listItems() {
        this.getItemsUseCase.execute(object : UseCase.UseCaseSubscriber<List<Item>>() {

            private val items = ArrayList<Item>()

            override fun onCompleted() {
                fragment.resetList(this.items)
            }

            override fun onNext(items: List<Item>) {
                this.items.addAll(items)
            }
        })
    }

    fun navigateItemDetail(item: Item) {
        val intent = CustomTabsIntent.Builder().setShowTitle(true).build()
        intent.launchUrl(fragment.activity, Uri.parse(item.url))
    }

}
